File: apps/android/app/src/main/java/org/llamenos/hotline/ui/cases/AssignmentViewModel.kt
package org.llamenos.hotline.ui.cases

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.llamenos.hotline.api.ApiService
import javax.inject.Inject

data class VolunteerSuggestion(
    val pubkey: String,
    val score: Int,
    val workloadScore: Int,
    val languageScore: Int,
    val specializationScore: Int,
    val availabilityScore: Int,
    val reasons: List<String>,
    val activeCaseCount: Int,
    val maxCases: Int,
    val matchedSpecializations: List<String>,
)

data class SuggestAssigneesResponse(
    val suggestions: List<VolunteerSuggestion>,
)

data class AssignmentUiState(
    val suggestions: List<VolunteerSuggestion> = emptyList(),
    val isLoading: Boolean = false,
    val isAssigning: Boolean = false,
    val error: String? = null,
    val assignedSuccess: Boolean = false,
)

@HiltViewModel
class AssignmentViewModel @Inject constructor(
    private val apiService: ApiService,
) : ViewModel() {

    private val _uiState = MutableStateFlow(AssignmentUiState())
    val uiState: StateFlow<AssignmentUiState> = _uiState.asStateFlow()

    fun loadSuggestions(recordId: String, language: String? = null) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                var path = apiService.hp("/api/records/$recordId/suggest-assignees")
                if (language != null) path += "?language=$language"
                val response = apiService.request<SuggestAssigneesResponse>("GET", path)
                _uiState.update { it.copy(suggestions = response.suggestions, isLoading = false) }
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }

    fun assign(recordId: String, pubkey: String, onSuccess: () -> Unit) {
        viewModelScope.launch {
            _uiState.update { it.copy(isAssigning = true, error = null) }
            try {
                apiService.request<Map<String, Any>>(
                    "POST",
                    apiService.hp("/api/records/$recordId/assign"),
                    body = mapOf("pubkeys" to listOf(pubkey)),
                )
                _uiState.update { it.copy(isAssigning = false, assignedSuccess = true) }
                onSuccess()
            } catch (e: Exception) {
                _uiState.update { it.copy(isAssigning = false, error = e.message) }
            }
        }
    }

    fun dismissError() {
        _uiState.update { it.copy(error = null) }
    }
}

File: apps/ios/Sources/ViewModels/AssignmentViewModel.swift
import Foundation

// MARK: - Volunteer suggestion model

struct VolunteerSuggestion: Decodable, Identifiable, Sendable {
    var id: String { pubkey }
    let pubkey: String
    let score: Int
    let workloadScore: Int
    let languageScore: Int
    let specializationScore: Int
    let availabilityScore: Int
    let reasons: [String]
    let activeCaseCount: Int
    let maxCases: Int
    let matchedSpecializations: [String]
}

struct AppSuggestAssigneesResponse: Decodable, Sendable {
    let suggestions: [VolunteerSuggestion]
}

// MARK: - AssignmentViewModel

@Observable
final class AssignmentViewModel {
    private let apiService: APIService

    var suggestions: [VolunteerSuggestion] = []
    var isLoading = false
    var isAssigning = false
    var errorMessage: String?

    init(apiService: APIService) {
        self.apiService = apiService
    }

    func loadSuggestions(for recordId: String, language: String? = nil) async {
        guard !isLoading else { return }
        isLoading = true
        errorMessage = nil
        do {
            var path = apiService.hp("/api/records/\(recordId)/suggest-assignees")
            if let lang = language {
                path += "?language=\(lang)"
            }
            let response: AppSuggestAssigneesResponse = try await apiService.request(method: "GET", path: path)
            suggestions = response.suggestions
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    func assign(recordId: String, pubkey: String) async -> Bool {
        isAssigning = true
        errorMessage = nil
        do {
            let body = ["pubkeys": [pubkey]]
            let _: CaseRecord = try await apiService.request(
                method: "POST",
                path: apiService.hp("/api/records/\(recordId)/assign"),
                body: body
            )
            isAssigning = false
            return true
        } catch {
            errorMessage = error.localizedDescription
            isAssigning = false
            return false
        }
    }
}
